upgrades/upgrade_testing_framework/cluster_management/cluster.py

```diff
@@ -0,0 +1,150 @@
+        return f"{self.name}-volume-{suffix}"
+
+    def start(self):
+        # TODO: Handle cluster already started
+
+        # Create network
+        network = self._docker_client.create_network(self._generate_network_name())
+        self._networks.append(network)
+
+        # Create shared volume(s)
+        # None for now
+
+        # Create any shared values
+        starting_master_nodes = [self._generate_node_name(node_num) for node_num in range(1, self._num_starting_nodes + 1)]
+        starting_seed_hosts = [self._generate_node_name(node_num) for node_num in range(1, self._num_starting_nodes + 1)]
+
+        for node_num in range(1, self._num_starting_nodes + 1):
+            # Create the node's data volume
+            node_volume = self._docker_client.create_volume(self._generate_volume_name(node_num))
+            self._volumes.append(node_volume)
+
+            # Create the node name, node config, container config
+            node_name = self._generate_node_name(node_num)
+            node_config = NodeConfiguration(node_name, self.name, starting_master_nodes, starting_seed_hosts)
+            port_mappings = [
+                PortMapping(f"920{node_num - 1}", "9200") # TODO: Breaks if more than 10 nodes (unlikely?)
+            ]
+            container_config = ContainerConfiguration(self._starting_image, network, port_mappings, self._volumes)
+
+            # Instantiate node
+            node = Node(node_name, container_config, node_config, self._docker_client)
+            self._nodes[node_name] = node
+
+            # Invoke start
+            node.start()
+
+    def wait_for_cluster_to_start_up(self, wait_limit_sec: int = None):
+        wait_interval_sec = 1 # time between checks, arbitrarily chosen
+        min_wait_time_sec = wait_interval_sec
+
+        if wait_limit_sec != None and wait_limit_sec < min_wait_time_sec:
+            raise ValueError("If you specify a wait time limit for cluster startup, it must be at least"
+                f" {min_wait_time_sec} seconds")
+
+        # Wait for the nodes in the cluster to be active; raise an exception if we exceed our time limit
+        nodes_to_wait_for = [node_name for node_name in self._nodes.keys()]
+        total_wait_time_sec = 0
+        while nodes_to_wait_for:
+            self.logger.debug(f"Waiting on the following nodes to be active: {nodes_to_wait_for}")
+            for node_name in nodes_to_wait_for:
+                if self._nodes[node_name].is_active():
+                    self.logger.info(f"Node {node_name} is now active")
+                    nodes_to_wait_for.remove(node_name)
+
+            if not nodes_to_wait_for: # all nodes active
+                break
+            
+            # TODO: Check resolution is wait_interval_sec, so might not respect wait_limit_sec if > 1
+            if not None == wait_limit_sec and total_wait_time_sec >= wait_limit_sec:
+                raise ClusterNotStartedInTimeException()
```

Comment:
Another thought here -- we need to deal with cleanup. presumably some of the nodes may have started successfully and need to be stopped, and all of them should be cleaned-up.

I'm fine with it being mostly out-of-scope, but maybe helpful to at least log a list of which ones did start and need to be stopped?

Reply:
All of the cluster state edge cases need more work.  I'm planning on iterating on those in the next pass (https://github.com/opensearch-project/opensearch-migrations/issues/7).  Mind waiting until the PR for that drops before looking at stuff like this too closely?